egui/src/widgets/plot/mod.rs

```diff
@@ -260,50 +275,104 @@ impl Widget for Plot {
-                auto_bounds: true,
+                auto_bounds: !min_auto_bounds.is_valid(),
+                hidden_curves: HashSet::new(),
             })
             .clone();
 
         let PlotMemory {
             mut bounds,
             mut auto_bounds,
+            mut hidden_curves,
         } = memory;
 
+        // Determine the size of the plot in the UI
         let size = {
-            let width = width.unwrap_or_else(|| {
-                if let (Some(height), Some(aspect)) = (height, view_aspect) {
-                    height * aspect
-                } else {
-                    ui.available_size_before_wrap_finite().x
-                }
-            });
-            let width = width.at_least(min_size.x);
-
-            let height = height.unwrap_or_else(|| {
-                if let Some(aspect) = view_aspect {
-                    width / aspect
-                } else {
-                    ui.available_size_before_wrap_finite().y
-                }
-            });
-            let height = height.at_least(min_size.y);
+            let width = width
+                .unwrap_or_else(|| {
+                    if let (Some(height), Some(aspect)) = (height, view_aspect) {
+                        height * aspect
+                    } else {
+                        ui.available_size_before_wrap_finite().x
+                    }
+                })
+                .at_least(min_size.x);
+
+            let height = height
+                .unwrap_or_else(|| {
+                    if let Some(aspect) = view_aspect {
+                        width / aspect
+                    } else {
+                        ui.available_size_before_wrap_finite().y
+                    }
+                })
+                .at_least(min_size.y);
             vec2(width, height)
         };
 
         let (rect, response) = ui.allocate_exact_size(size, Sense::drag());
+        let plot_painter = ui.painter().sub_region(rect);
 
         // Background
-        ui.painter().add(Shape::Rect {
+        plot_painter.add(Shape::Rect {
             rect,
             corner_radius: 2.0,
             fill: ui.visuals().extreme_bg_color,
             stroke: ui.visuals().window_stroke(),
         });
 
+        // --- Legend ---
+
+        if show_legend {
+            // Collect the legend entries.
+            let mut legend_entries: Vec<_> = curves
+                .iter()
+                .filter(|curve| !curve.name.is_empty())
+                .map(|curve| {
+                    let checked = !hidden_curves.contains(&curve.name);
+                    let text = curve.name.clone();
+                    let color = curve.stroke.color;
+                    LegendEntry::new(text, color, checked)
+                })
+                .collect();
+
+            // Show the legend.
+            if !legend_entries.is_empty() {
+                let legend = PlotLegend::new(&mut legend_entries, rect);
+                let response = legend.ui(ui);
```

Comment:
Instead of doing manual placement of the legends, try this:

``` rust
// Put legend on top of plot:
let mut legend_ui = ui.child_ui(rect, Layout::top_down(Align::LEFT)); 
for entry in &mut legend_entries {
    if legend_ui.add(entry).hovered() {
        show_x = false;
        show_y = false;
    }
}
```

with 


``` rust
impl Widget for &mut LegendEntry {
    fn ui(self, ui: &mut Ui) -> Response {
        // same as before but WITHOUT `ui.with_layout(Layout::top_down(Align::LEFT), |ui| {`
    }
}
```


No need for `struct PlotLegend` or `struct LegendEntryWidget` !

Reply:
Ha! So good, I knew there would be a better way but had no idea that it would be _that_ elegant.